test/test_dynamic_shapes.py

```diff
@@ -816,5 +833,944 @@ def is_complex(x):
                 self.assertEqual(op.is_integer, None)
                 self.assertTrue(op.is_real)
 
+
+class TestDimConstraints(TestCase):
+    def test_dim_constraints_reduce_congruences_simple(self):
+        from sympy import Symbol
+        from torch.fx.experimental.symbolic_shapes import DimConstraints
+
+        s = Symbol("s", positive=True, integer=True)
+        dim_constraints = DimConstraints({}, {})
+        dim_constraints._congruences[s] = {
+            (s / 2) % 2,
+            (s / 2) % 8,
+            (s / 2) % 4,
+            s % 2,
+            ((s / 16) + 2) % 4,
+        }
+        congruences = dim_constraints.reduce_congruences()
+        self.assertEqual(congruences[s], {(s + 32) % 64})
+
+    def test_dim_constraints_reduce_inequalities_simple(self):
+        from sympy import Eq, Interval, Ne, Symbol
+        from sympy.solvers.inequalities import reduce_inequalities
+
+        s = Symbol("s", positive=True, integer=True)
+        exprs = {
+            s >= 2,
+            Ne(8 * s, 16),
+            Ne(s / 2, 1),
+            Ne(16 * s, 32),
+            s < 16,
+            Ne(s, 2),
+            s / 2 < 16,
+            s / 2 > 1,
+            s / 2 >= 2,
+            Ne(3 * s / 2, 3),
+        }
+        solution = reduce_inequalities(exprs, s).as_set()
+        self.assertEqual(solution, Interval.Ropen(4, 16))
+
+        exprs.add(Eq(s / 2, 4))
+        solution = reduce_inequalities(exprs, s).as_set()
+        self.assertEqual(solution, {8})
+
+    def test_dim_constraints_solve_full(self):
```

Comment:
These constraints were lifted from somewhere, right? Seems worth mentioning where you got it from.